src/query/api/v1/handler/prometheus/native/common.go

```diff
@@ -322,6 +334,20 @@ func renderResultsJSON(
 			jw.BeginArray()
 			jw.WriteInt(int(dp.Timestamp.Unix()))
 			jw.WriteString(utils.FormatFloat(dp.Value))
+			if includeExemplars && examplarIdx < len(exemplars) {
+				if exemplars[examplarIdx].Timestamp.Unix() <= dp.Timestamp.Unix() {
+					for {
+						i := examplarIdx + 1
+						if i < len(exemplars) && exemplars[i].Timestamp.Unix() <= dp.Timestamp.Unix() {
```

Comment:
Also we need to check if the exemplar timestamp is not older than `dp.Timestamp - resolution`

Reply:
Might be good to move this logic into a function? LIke `alignExemplars(result ServeResult) ServeResult`? Then it should be easy to remove from heret when we do the more indepth solution